src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserRoleService.cs

```diff
@@ -114,12 +121,114 @@ public class UserRoleService : IUserRoleService
+
                                 return _emailTemplateService.SendReleaseRoleEmail(user.Email, release, role);
                             });
                     })
                 );
         }
 
+        private async Task SetExclusiveGlobalRole(string? globalRoleNameToSet, ApplicationUser user)
+        {
+            var existingRoleNames = await _identityUserManager
+                .GetRolesAsync(user) ?? new List<string>();
+
+            if (globalRoleNameToSet == null)
+            {
+                await _identityUserManager.RemoveFromRolesAsync(user, existingRoleNames);
+                return;
+            }
+
+            if (!existingRoleNames.Contains(globalRoleNameToSet))
+            {
+                await _identityUserManager.AddToRoleAsync(user, globalRoleNameToSet);
+            }
+
+            var rolesToRemove = existingRoleNames
+                .Where(roleName => roleName != globalRoleNameToSet)
+                .ToList();
+
+            if (rolesToRemove.Count > 0)
+            {
+                await _identityUserManager.RemoveFromRolesAsync(user, rolesToRemove);
+            }
+        }
+
+        private async Task UpgradeToGlobalRoleIfRequired(string globalRoleNameToSet, ApplicationUser user)
+        {
+            var existingRoleNames = await _identityUserManager
+                .GetRolesAsync(user) ?? new List<string>();
+
+            var userAlreadyAssignedToRole = existingRoleNames.Contains(globalRoleNameToSet);
+            
+            var higherRoleAlreadyAssigned = !existingRoleNames
+                .Intersect(GetHigherRoles(globalRoleNameToSet))
+                .IsNullOrEmpty();
+
+            if (!userAlreadyAssignedToRole && !higherRoleAlreadyAssigned)
+            {
+                await _identityUserManager.AddToRoleAsync(user, globalRoleNameToSet);
+            }
+
+            var lowerRolesToRemove = GetLowerRoles(globalRoleNameToSet)
+                .Intersect(existingRoleNames)
+                .ToList();
+
+            if (lowerRolesToRemove.Count > 0)
+            {
+                await _identityUserManager.RemoveFromRolesAsync(user, lowerRolesToRemove);
+            }
+        }
+        
+        private async Task DowngradeFromGlobalRoleIfRequired(ApplicationUser user, string globalRoleNameToDowngradeFrom)
+        {
+            var existingGlobalRoleNames = await _identityUserManager
+                .GetRolesAsync(user) ?? new List<string>();
+
+            var higherPrecedenceExistingGlobalRoleNames = existingGlobalRoleNames
+                .Where(role => GlobalRolePrecedenceOrder.IndexOf(role) >
+                    GlobalRolePrecedenceOrder.IndexOf(globalRoleNameToDowngradeFrom));
+            
+            var requiredGlobalRoleNames = 
+                await GetRequiredGlobalRoleNamesForResourceRoles(user);
+
+            var highestPrecedenceRoleNameToRetain = higherPrecedenceExistingGlobalRoleNames
```

Comment:
Yeah it works at the moment with the current model but can imagine in the future it might not.  I'd think that at that point we might then look at doing the bigger job of removing the Global Roles altogether, which we decided not to do at this point.